### PR TITLE
Add new files to the folder

### PR DESCRIPTION
**Pull request by Tashwin**

### CODE REVIEW

**Comment by caguero** [comment]
Running the example.py I got this error:

``` python
Traceback (most recent call last):
  File "example.py", line 7, in <module>
    roadPointWidthMap = osmRoads.getRoadCoords()
AttributeError: Osm2Dict instance has no attribute 'getRoadCoords'
```

It seems that the class Osm2Dict does not have a getRoadCoords() method.

**Comment by caguero** [comment]
I recommend to follow [pep8 style guide](http://www.python.org/dev/peps/pep-0008/). What I normally do is install a pep8 linter that automatically hightlights all the instructions that are not pep8 compliant. Here is [an example](https://github.com/SublimeLinter/SublimeLinter-pep8) for SublitmeText editor.

**Comment by Tashwin** [comment]
I just added example.py to the git account. This is updated and tested.

**Comment by Tashwin** [comment]
Thanks I'll look it up and install pep8 linter

**Comment by wjwwood** [comment]
Also, you can setup an automated test for pep8, like I did for bloom:

https://github.com/ros-infrastructure/bloom/blob/master/test/test_code_format.py

**Comment by Tashwin** [comment]
Thanks! that helped a lot ! 
I'll make the files pep8 compliant and make a fresh commit.

**Comment by caguero** [comment]
Not a big deal but if you manage to keep all the lines with a maximum of 79 chars, all your files will have zero pep8 warnings/errors.

I see a sampleMapImage.png file. That should be removed from the repo. It will be generated on demand by your script.

**Comment by Tashwin** [comment]
I just deleted the sampleMapImage.png files. I will also update the other files and commit the changes to the repo.

File: osm2dict.py [hunk 1]
```diff
@@ -0,0 +1,281 @@
...
+
+        while 1:
+            #get map coords
+            print("\nPlease enter the latitudnal and logitudnal" +
+                  "coordinates of the area or select from default by hitting return twice \n")
+
+            startCoords = raw_input("Enter starting coordinates: [lon lat] :").split(' ')
+            endCoords = raw_input("Enter ending coordnates: [lon lat]: ").split(' ')
+
+            if startCoords != [] and endCoords != [] and len(startCoords) == 2 and len(endCoords) == 2:
```
Comment (caguero) [inline osm2dict.py]: You could type:

``` python
  if startCoords
```

instead of:

``` python
 if startCoords != []
```

File: osm2dict.py [hunk 2]
```diff
@@ -0,0 +1,281 @@
+#!/usr/bin/env python
+import osmapi
+import numpy as np
+from getMapImage import getMapImage
+
+
+class Osm2Dict:
+
+    def __init__(self):
```
Comment (caguero) [inline osm2dict.py]: I'd suggest to add latStart, lonStart, latStop and lonStop as parameters of the constructor. You could move the interaction with the user to a main program.

File: osm2dict.py [hunk 3]
```diff
@@ -0,0 +1,281 @@
...
+        angles = self.latLongBearing(coords)
+
+        point = np.array([distance*np.cos(angles)*1000,
+                          distance*np.sin(angles)*1000,
+                          np.zeros(np.shape(distance))*1000])
+
+        return point
+
+    def getMapDetails(self):
+        ''' Returns a list of highways with corresponding widhts and a list of all the models to be included'''
```
Comment (caguero) [inline osm2dict.py]: Typo: widhts <--> widths

**Comment by caguero** [comment]
We could remove sampleOutput.sdf too, if is generated by a script

**Comment by caguero** [comment]
It would be nice if every class has a test that checks its correct operation. Take a look at a [unit test module](http://docs.python.org/2/library/unittest.html) for an ever better approach.

**Comment by Tashwin** [comment]
Yes, that was my next question. If I should make the programs, command line compatible for input arguments. 
Next, for now I simply included the image generation as part of the osm2dict program, is that fine?
Also, I will look into the unit testing modules.

**Comment by caguero** [comment]
Yes, the image will be generated in your osm2dict, that's perfect. 

Try to isolate the class that implements some functionality from the program that tests it. The program that tests your class/es might be interactive (as it is right now) or could use command line arguments. If you want to add support for command line arguments take a look at [argparse module](http://docs.python.org/dev/library/argparse.html).

**Comment by Tashwin** [comment]
Ok. I'll incorporate all the changes. 
So, I will get rid of all the interaction happening from osm2dict, and do it from a main test program like example.py.
I have used argparse module before, so I'm familiar with its working and can incorporate it in  our test program or keep it  somewhat similar to as it is now. So, I think I'll I push this back and work on building unit tests first?

**Comment by caguero** [comment]
Sounds good.

**Comment by scpeters** [comment]
Regarding the image generation in osm2dict, I think it might be better to put it in a separate file, in order to separate the dependencies. If someone just wants to get osm data, they have to install mapnik if they're both in the same file.

**Comment by Tashwin** [comment]
Ok that has a file for itself. I can do that simply enough.

**Comment by Tashwin** [comment]
I added a new file which goes in and downloads the actual osm data, which is then used by both, the getMapImage.py and osm2dict.py. This prevents downloading same data twice.
Should I add a clean-up function which deletes the image and .osm files created during the build?

File: getMapImage.py [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+        text_symbolizer.halo_fill = mapnik.Color('white')
+
+        rule2.symbols.append(text_symbolizer)
+
+        styleType.rules.append(rule)
+        styleType.rules.append(rule2)
+
+        m.append_style(highwayType, styleType)
+
+    ds = mapnik.Osm(file = osmFile)
```
Comment (caguero) [inline getMapImage.py]: Remove spaces around parameter equals

File: getOsmFile.py [hunk 5]
```diff
@@ -0,0 +1,24 @@
+import urllib2
+import osmapi
+
+
+def getOsmFile(box, filename = 'map.osm'):
```
Comment (caguero) [inline getOsmFile.py]: Remove spaces around parameter equals

File: getOsmFile.py [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+
+    osmFile = urllib2.urlopen('http://api.openstreetmap.org/api/0.6/map?bbox='
+                              + str(box)[1:-1].replace(" ", ""))
+
+    osm = open('map.osm','w')
```
Comment (caguero) [inline getOsmFile.py]: Add space after the comma

File: osm2dict.py [hunk 7]
```diff
@@ -0,0 +1,308 @@
...
+
+                    if typeHighway in self.highwayType.keys():
+
+                                roadName = tagData.get("name")
+
+                                if roadName is None:
+                                    roadName = (typeHighway +
+                                                "_" +
+                                                str(self.data[i].get("data")
+                                                           .get("id")))
```
Comment (caguero) [inline osm2dict.py]: Remove one or two tabs

File: osm2dict.py [hunk 8]
```diff
@@ -0,0 +1,308 @@
...
+                    typeHighway = tagData.get("highway")
+
+                    if typeHighway in self.highwayType.keys():
+
+                                roadName = tagData.get("name")
+
+                                if roadName is None:
+                                    roadName = (typeHighway +
+                                                "_" +
+                                                str(self.data[i].get("data")
```
Comment (caguero) [inline osm2dict.py]: That line should be <= 79 chars

File: osm2dict.py [hunk 9]
```diff
@@ -0,0 +1,308 @@
...
+                                                                 .get("id"))
+
+                                node_ref = self.data[i].get("data").get("nd")
+                                coords = np.array([])
+                                for j in range(len(self.data)):
+
+                                    if "node" in self.data[j].get("type"):
+
+                                        if (self.data[j].get("data")
+                                                   .get("id")) in node_ref:
```
Comment (caguero) [inline osm2dict.py]: Remove one tab

File: osm2dict.py [hunk 10]
```diff
@@ -0,0 +1,308 @@
...
+                                buildingName += tagData.get("name_1")
+
+                        node_ref = self.data[i].get("data").get("nd")
+                        coords = np.array([])
+
+                        for j in range(len(self.data)):
+
+                            if "node" in self.data[j].get("type"):
+
+                                if self.data[j].get("data").get("id") in node_ref:
```
Comment (caguero) [inline osm2dict.py]: That line should be <= 79 chars

File: osm2dict.py [hunk 11]
```diff
@@ -0,0 +1,308 @@
...
+
+                                    coords = np.append(coords,
+                                                       self.data[j].get("data")
+                                                              .get("lon"))
```
Comment (caguero) [inline osm2dict.py]: Line over-indented. My pep8 checker complains here.

File: osm2dict.py [hunk 12]
```diff
@@ -0,0 +1,308 @@
...
+                                    coords = np.append(coords,
+                                                       self.data[j].get("data")
+                                                              .get("lat"))
```
Comment (caguero) [inline osm2dict.py]: Line over-indented. My pep8 checker complains here.

File: osm2dict.py [hunk 13]
```diff
@@ -0,0 +1,308 @@
...
+                        amenity = tagData.get("amenity")
+
+                        node_ref = self.data[i].get("data").get("nd")
+                        coords = np.array([])
+
+                        for j in range(len(self.data)):
+
+                            if "node" in self.data[j].get("type"):
+
+                                if self.data[j].get("data").get("id") in node_ref:
```
Comment (caguero) [inline osm2dict.py]: Line too long.

File: osm2dict.py [hunk 14]
```diff
@@ -0,0 +1,308 @@
...
+
+                                    coords = np.append(coords,
+                                                       self.data[j].get("data")
+                                                              .get("lon"))
```
Comment (caguero) [inline osm2dict.py]: Line over-indented. My pep8 checker complains here.

File: osm2dict.py [hunk 15]
```diff
@@ -0,0 +1,308 @@
...
+                                    coords = np.append(coords,
+                                                       self.data[j].get("data")
+                                                              .get("lat"))
```
Comment (caguero) [inline osm2dict.py]: Line over-indented. My pep8 checker complains here.

**Comment by caguero** [comment]
I've added some minor pep8 comments.

**Comment by caguero** [comment]
Related with the command line mode. I'm envisionning something like this:
gz_osm [options] <startLat> <startLong> <endLat> <endLong>

Options:
 -v [ --version ] Output version information.
 -h [--help]       Produce this help message.
 -a [--all]          Import all the objects.
 -b [--buildings] Import buildings.
 -d [--destination] Set the destination directory where the .sdf file will be generated.
 -m [--map]      Generate a .png file with the map.
 -r [--roads]      Import roads.
 -s [--sidewalks] Import sidewalks.
 -t [--traffic-lights] Import traffic lights.

**Comment by Tashwin** [comment]
Thanks. I fixed the pep8 inconsistencies, added a few test files. 
I am currently working on the argparse functionality.

**Comment by scpeters** [comment]
I'm having trouble running the gz_osm.py command with no arguments (which should maybe return a help string?):

```
$ ./gz_osm.py 
Traceback (most recent call last):
  File "./gz_osm.py", line 45, in <module>
    getMapImage(args.osmFile, args.imageFile)
TypeError: getMapImage() takes exactly 1 argument (2 given)
```

**Comment by Tashwin** [comment]
I forgot to update the getMapImage.py. It should work now.

**Comment by Tashwin** [comment]
I  have added some new features to the maps.
The sdf file which is generated works well for smaller areas, but for larger areas, I am getting the following error from gazebo.

```
$ gazebo outFile.sdf 
Gazebo multi-robot simulator, version 2.2.1
Copyright (C) 2012-2013 Open Source Robotics Foundation.
Released under the Apache 2 License.
http://gazebosim.org

Msg Waiting for master
Msg Connected to gazebo master @ http://127.0.0.1:11345
Msg Publicized address: 192.168.1.62
Gazebo multi-robot simulator, version 2.2.1
Copyright (C) 2012-2013 Open Source Robotics Foundation.
Released under the Apache 2 License.
http://gazebosim.org

Msg Waiting for master
Msg Connected to gazebo master @ http://127.0.0.1:11345
Msg Publicized address: 192.168.1.62
Error [ConnectionManager.cc:565] Unable to get namespaces from master
Warning [Publisher.cc:134] Queue limit reached for topic /gazebo/default/model/info, deleting message. This warning is printed only once.
```

I have run gzsdf check <filename> which returns a 'Success' value.

Could you point me as to where I might be going wrong?

**Comment by caguero** [comment]
:+1:  We should start thinking merging this pull request

**Comment by Tashwin** [comment]
Yes, I added the comments to the files. Also, added support for a directly input osm file. I am working on the footway problem and if I can't solve it by tomorrow I'll create an issue regarding it.

**Comment by scpeters** [comment]
Yes, let's merge and then use the issue tracker for future improvements.

**Pull request merged**